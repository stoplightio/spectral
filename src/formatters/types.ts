import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
