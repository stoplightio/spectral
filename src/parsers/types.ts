import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
