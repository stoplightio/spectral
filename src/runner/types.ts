import type { DocumentInventory } from '../documentInventory';
import type { Ruleset } from '../ruleset/ruleset';
import { ISpectralDiagnostic } from '../types';

export interface IRunnerInternalContext {
  ruleset: Ruleset;
  documentInventory: DocumentInventory;
  results: ISpectralDiagnostic[];
  promises: Array<Promise<void>>;
}
