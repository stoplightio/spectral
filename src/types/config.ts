import { HumanReadableDiagnosticSeverity } from './rule';

export type FailSeverity = HumanReadableDiagnosticSeverity;

export enum OutputFormat {
  JSON = 'json',
  STYLISH = 'stylish',
  JUNIT = 'junit',
  HTML = 'html',
  TEXT = 'text',
  TEAMCITY = 'teamcity',
}

export interface ILintConfig {
  encoding: string;
  format: OutputFormat;
  output?: string;
  resolver?: string;
  ruleset?: string[];
  skipRule?: string[];
  ignoreUnknownFormat: boolean;
  showUnmatchedGlobs: boolean;
  failOnUnmatchedGlobs: boolean;
  verbose?: boolean;
  quiet?: boolean;
}
