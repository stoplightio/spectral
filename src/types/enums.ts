export enum RuleType {
  VALIDATION = 'validation',
  STYLE = 'style',
}

export enum RuleFunction {
  ALPHABETICAL = 'alphabetical',
  CASING = 'casing',
  LENGTH = 'length',
  PATTERN = 'pattern',
  ENUM = 'enumeration',
  SCHEMA = 'schema',
  SCHEMAPATH = 'schemaPath',
  TRUTHY = 'truthy',
  XOR = 'xor',
  TYPED_ENUM = 'typedEnum',
}
