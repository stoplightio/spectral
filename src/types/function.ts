import { JsonPath } from '@stoplight/types';
import { DocumentInventory } from '../documentInventory';

export type IFunction<O = any> = (
  targetValue: any,
  options: O,
  paths: IFunctionPaths,
  otherValues: IFunctionValues,
) => void | IFunctionResult[];

export interface IFunctionPaths {
  given: JsonPath;
  target?: JsonPath;
}

export interface IFunctionValues {
  original: any;
  given: any;
  documentInventory: DocumentInventory;
}

export interface IFunctionResult {
  message: string;
  path?: JsonPath;
}
