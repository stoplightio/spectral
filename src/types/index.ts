export * from './spectral';
export * from './function';
