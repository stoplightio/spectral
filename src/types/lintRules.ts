import { IRuleDefinitionBase } from './index';

export interface IRuleParam {
  properties: string | string[];
}

export interface IRuleStringParam extends IRuleParam {
  value: string;
}

export interface IRuleNumberParam {
  value: number;
  property?: string;
}

export interface IAlphaRuleParam extends IRuleParam {
  // if sorting objects, use key for comparison
  keyedBy?: string;
}

export interface IRulePatternParam {
  // object key to apply rule to
  property: string;

  // value to use for rule
  value: string;

  // value to omit from regex matching
  omit?: string;

  // value to split the property on prior to performing regex matching
  split?: string;
}

export interface ITruthyRule extends IRuleDefinitionBase {
  type: 'truthy';

  // key(s) of object that should evaluate as 'truthy' (considered true in a
  // boolean context)
  truthy: string | string[];

  properties?: number;
}

export interface IOrRule extends IRuleDefinitionBase {
  type: 'or';

  // test to verify if any of the provided keys are present in object
  or: string[];
}

export interface IXorRule extends IRuleDefinitionBase {
  type: 'xor';

  // test to verify if one (but not all) of the provided keys are present in
  // object
  xor: string[];
}

export interface IMaxLengthRule extends IRuleDefinitionBase {
  type: 'maxLength';

  // verify property is under a specified number of characters
  maxLength: IRuleNumberParam;
}

export interface IAlphaRule extends IRuleDefinitionBase {
  type: 'alphabetical';

  // verify property is within alphabetical order
  alphabetical: IAlphaRuleParam;
}

export interface INotEndWithRule extends IRuleDefinitionBase {
  type: 'notEndWith';

  // verify property does not end with string
  notEndWith: IRulePatternParam;
}

export interface INotContainRule extends IRuleDefinitionBase {
  type: 'notContain';

  // verify property does not contain value
  notContain: IRuleStringParam;
}

export interface IPatternRule extends IRuleDefinitionBase {
  type: 'pattern';

  // run regex match
  pattern: IRulePatternParam;
}

export interface IFunctionRule extends IRuleDefinitionBase {
  type: 'function';
  function: string; // eventually allow js functions here? or references to one
  args?: any; //need to think about this a little
}

export type LintRule =
  | ITruthyRule
  | IOrRule
  | IXorRule
  | IMaxLengthRule
  | IAlphaRule
  | INotEndWithRule
  | INotContainRule
  | IPatternRule
  | IFunctionRule;
