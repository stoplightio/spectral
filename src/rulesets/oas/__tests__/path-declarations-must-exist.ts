import { DiagnosticSeverity } from '@stoplight/types';
import type { Spectral } from '../../../spectral';
import { createWithRules } from './__helpers__/createWithRules';

describe('path-declarations-must-exist', () => {
  let s: Spectral;

  beforeEach(async () => {
    s = await createWithRules(['path-declarations-must-exist']);
  });

  test('validate a correct object', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: { '/path/{parameter}': {} },
    });
    expect(results.length).toEqual(0);
  });

  test('return errors if parameter is empty', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: { '/path/{}': {} },
    });
    expect(results).toEqual([
      {
        code: 'path-declarations-must-exist',
        message: 'Path parameter declarations cannot be empty, ex.`/given/{}` is invalid.',
        path: ['paths', '/path/{}'],
        range: {
          end: {
            character: 18,
            line: 3,
          },
          start: {
            character: 15,
            line: 3,
          },
        },
        severity: DiagnosticSeverity.Warning,
      },
    ]);
  });
});
