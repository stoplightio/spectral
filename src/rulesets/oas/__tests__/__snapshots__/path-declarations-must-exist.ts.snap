// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path-declarations-must-exist return errors if parameter is empty 1`] = `
Array [
  Object {
    "code": "path-declarations-must-exist",
    "message": "given declarations cannot be empty, ex.\`/given/{}\` is invalid.",
    "path": Array [
      "paths",
      "/path/{}",
    ],
    "range": Object {
      "end": Object {
        "character": 18,
        "line": 3,
      },
      "start": Object {
        "character": 15,
        "line": 3,
      },
    },
    "severity": 1,
    "source": undefined,
    "summary": "given declarations cannot be empty, ex.\`/given/{}\` is invalid.",
  },
]
`;
